### ISSUE TITLE
Non-Ecore EDataTypes and oclIsKindOf do not work

### ISSUE THREAD

**Issue by bluezio** [atlanmod/Mogwai#4]
I have set up a small prototype that loads the GraBaTs'09 `set0.xmi` into NeoEMF and then runs a slightly more detailed query than the one in this project's benchmarks:

```
import DOM : 'platform:/resource/jdtast.neoemf/model/JDTAST.ecore'

package DOM

context TypeDeclaration

def: singletons : Set(TypeDeclaration) =

  TypeDeclaration.allInstances()
  ->select(td |
    td.bodyDeclarations
      ->exists(md |
      	md.oclIsTypeOf(MethodDeclaration)
      	-- this cannot be translated by Mogwai, apparently
        and md.modifiers->exists(mod | mod.oclIsKindOf(Modifier) and mod.oclAsType(Modifier).public)
        and md.modifiers->exists(mod | mod.oclIsKindOf(Modifier) and mod.oclAsType(Modifier)._static)
        and md.oclAsType(MethodDeclaration).returnType.oclIsTypeOf(SimpleType)
        -- and this doesn't work, either - fullyQualifiedName does not appear in the translated query
        and md.oclAsType(MethodDeclaration).returnType.oclAsType(SimpleType).name.fullyQualifiedName = td.name.fullyQualifiedName
      )
  )

endpackage
```

Sadly, it doesn't seem to be translated correctly: see (1), (2) and (3).

```
def metaTypeDeclaration = g.getIndex("metaclasses",Vertex.class)[[name:"TypeDeclaration"]];
def metaMethodDeclaration = g.getIndex("metaclasses",Vertex.class)[[name:"MethodDeclaration"]];
def metaModifier = g.getIndex("metaclasses",Vertex.class)[[name:"Modifier"]];
def metaSimpleType = g.getIndex("metaclasses",Vertex.class)[[name:"SimpleType"]];
def metaTypeDeclarationNode = metaTypeDeclaration.next();
def metaMethodDeclarationNode = metaMethodDeclaration.next();
def metaModifierNode = metaModifier.next();
def metaSimpleTypeNode = metaSimpleType.next();

metaTypeDeclarationNode.inE("kyanosInstanceOf").outV.filter{
  def td = it;
  td.outE("bodyDeclarations").inV.filter{
    def md = it;
    md.transform {
      it.outE("kyanosInstanceOf").inV.next() == metaMethodDeclarationNode;}.next()
      && md.outE("modifiers").inV.filter{
        def mod = it; mod && mod.; -- (1)
      }.hasNext()
     && md.outE("modifiers").inV.filter{
       def mod = it;mod && mod.; -- (2)
     }.hasNext()
     && md.outE("returnType").inV.transform{
      it.outE("kyanosInstanceOf").inV.next() == metaSimpleTypeNode;
    }.next()
    && md.outE("returnType").inV.outE("name").inV == td.outE("name").inV; -- (3)
}.hasNext();
};
```

It feels as if the `oclAsType(...)` calls (required by the OCL parser) were preventing the translator from working correctly. Do you think this would be the case?

**Comment by bluezio** [comment]
I tried turning oclAsType(...) calls into Gremlin identity steps with this extra rule, but that didn't change things as much as I expected:

```
-- Transforms an oclAsType operation into equivalent Gremlin operation.
-- Gremlin is dynamically typed, so this can be left as an identity step.
rule oclAsType2steps {
	from
		oclAsType : OCL!OperationCallExp (oclAsType.getOpName() = 'oclAsType')
	to
		lastColStep : Gremlin!IdentityStep (
			nextElement <- oclAsType.getComposite()
		)
}
```

The generated query is now as follows:

```
def metaTypeDeclaration = g.getIndex("metaclasses",Vertex.class)[[name:"TypeDeclaration"]];
def metaMethodDeclaration = g.getIndex("metaclasses",Vertex.class)[[name:"MethodDeclaration"]];
def metaModifier = g.getIndex("metaclasses",Vertex.class)[[name:"Modifier"]];
def metaSimpleType = g.getIndex("metaclasses",Vertex.class)[[name:"SimpleType"]];
def metaTypeDeclarationNode = metaTypeDeclaration.next();
def metaMethodDeclarationNode = metaMethodDeclaration.next();
def metaModifierNode = metaModifier.next();
def metaSimpleTypeNode = metaSimpleType.next();
metaTypeDeclarationNode.inE("kyanosInstanceOf").outV.filter{def td = it;td.outE("bodyDeclarations").inV.filter{def md = it;md.transform{it.outE("kyanosInstanceOf").inV.next() == metaMethodDeclarationNode;}.next() && md.outE("modifiers").inV.filter{def mod = it;mod && mod._();}.hasNext() && md.outE("modifiers").inV.filter{def mod = it;mod && mod._();}.hasNext() && md._().outE("returnType").inV.transform{it.outE("kyanosInstanceOf").inV.next() == metaSimpleTypeNode;}.next() && md._().outE("returnType").inV._().outE("name").inV == td.outE("name").inV;}.hasNext();};
```

**Comment by bluezio** [comment]
It seems oclAsType(...) is supported, but the problem lies rather with the varExp2steps. I tried this simpler query, and it wouldn't translate well into Gremlin:

```
import DOM : 'platform:/resource/jdtast.neoemf/model/JDTAST.ecore'

package DOM

context TypeDeclaration

def: public : Set(Modifier) =
  Modifier.allInstances()->select(m | m.public = true)

endpackage
```

It complains about this:
```
Cannot set feature nextElement to value m.public, inter-model references are forbidden. Configure launching options to allow them.
```

That seems to come from line 322 of ocl2gremlin.atl, which is within this rule:
```
rule varExp2steps {
	from
		varExp : OCL!VariableExp(
			not(varExp.isInCollect()) or varExp.referredVariable.name='self'
		)
	to
		va : Gremlin!VariableAccess(
			name <- varExp.referredVariable.name.debug('varExp2steps::name'),
			nextElement <- varExp.getComposite().debug('varExp2steps::nextElement')
		)
}
```

**Comment by bluezio** [comment]
I have investigated a bit more: it seems the problem was that the JDTAST metamodel I'm using has a few custom EDataTypes based on EString/EInt/EBoolean, which are named String, Integer and Boolean. Tweaking `propertyAttStringLastInComparison2steps` and other similar rules to accept these solved some of the issues.

Another problem was that `oclIsKindOf` is apparently not supported, so I replaced it with `oclIsTypeOf`. I understand that this is not directly supported since NeoEMF doesn't add isKindOf edges, only isInstanceOf edges, and types are not related. Would this be correct?

**Comment by bluezio** [comment]
I've created a PR for the fix for non-Ecore EDataTypes.

**Comment by gdaniel** [comment]
> Another problem was that oclIsKindOf is apparently not supported, so I replaced it with oclIsTypeOf. I understand that this is not directly supported since NeoEMF doesn't add isKindOf edges, only isInstanceOf edges, and types are not related. Would this be correct?

Yes, the "kindOf" relationship is not explicitly represented in NeoEMF, making its translation to Gremlin more complicated. However I think it is possible by inspecting the input metamodel, I'll do some tests to see what I can do on that.

**Comment by gdaniel** [comment]
> I have investigated a bit more: it seems the problem was that the JDTAST metamodel I'm using has a few custom EDataTypes based on EString/EInt/EBoolean, which are named String, Integer and Boolean. Tweaking propertyAttStringLastInComparison2steps and other similar rules to accept these solved some of the issues.

Did it solve the problem in this query you linked?
`import DOM : 'platform:/resource/jdtast.neoemf/model/JDTAST.ecore'

package DOM

context TypeDeclaration

def: public : Set(Modifier) =
  Modifier.allInstances()->select(m | m.public = true)

endpackage`

I have tested it with my JDTAST metamodel and the translation is correct with the fix you provide in #5

**Comment by gdaniel** [comment]
Looking at the `oclIsKindOf `operation, I will push a quick fix for that: now the translation inspects the input metamodel to find all the subclasses of EClasses involved in `oclIsKindOf `operations and use them to provide the correct translation to Gremlin. However there is a similar error for `allInstances()` operations: in the current version it only returns strict instances where it should also include instances of subclasses.

**Comment by gdaniel** [comment]
Fixed in 2049a2d6e304c840dbb7a07a08714e883bc14575,` oclIsKindOf(Class)` is now supported using input metamodel inspection to find all classes that are subclasses of `Class`

Non-Ecore datatypes error was solved in your PR (#5)
